fsthttp/cache.go

```diff
@@ -0,0 +1,761 @@
+		if keys != "" {
+			keys += " "
+		}
+		keys += extra
+	}
+
+	const fastlyHeaderSurrogateKey = "surrogate-key"
+
+	vals := candidateResponse.abiResp.GetHeaderValues(fastlyHeaderSurrogateKey, RequestLimits.maxHeaderNameLen)
+	for vals.Next() {
+		if keys != "" {
+			keys += " "
+		}
+		keys += string(vals.Bytes())
+	}
+	if err := vals.Err(); err != nil {
+		return nil, fmt.Errorf("get header values: %w", err)
+	}
+	opts.surrogate = keys
+
+	return opts, nil
+}
+
+func httpCacheGetSuggestedCacheWriteOptions(cacheHandle *fastly.HTTPCacheHandle, resp *fastly.HTTPResponse) (*cacheWriteOptions, error) {
+	var opts cacheWriteOptions
+	opts.abiOpts.FillConfigMask()
+
+	newABIOpts, err := fastly.HTTPCacheGetSuggestedCacheOptions(cacheHandle, resp, &opts.abiOpts)
+	if err != nil {
+		return nil, fmt.Errorf("get suggested cache options: %w", err)
+	}
+
+	newOpts := &cacheWriteOptions{abiOpts: *newABIOpts}
+	newOpts.loadFromABI()
+
+	return newOpts, nil
+}
+
+func (candidateResponse *CandidateResponse) finalizeOptions() (fastly.HTTPCacheStorageAction, *cacheWriteOptions, error) {
+	var storageAction = candidateResponse.suggestedStorageAction
+	if candidateResponse.useStorageAction {
+		storageAction = candidateResponse.overrideStorageAction
+	}
+
+	suggestedCacheWriteOptions := candidateResponse.suggestedCacheWriteOptions
+	if suggestedCacheWriteOptions == nil {
+		var err error
+		suggestedCacheWriteOptions, err = candidateResponse.buildFreshSuggestedCacheWriteOptions()
+		if err != nil {
+			return 0, nil, err
+		}
+	}
+
+	var opts cacheWriteOptions
+
+	if candidateResponse.useTTL {
+		opts.maxAge = candidateResponse.overrideTTL - suggestedCacheWriteOptions.age
+	} else {
+		opts.maxAge = suggestedCacheWriteOptions.maxAge
+	}
+
+	opts.age = suggestedCacheWriteOptions.age
+
+	if candidateResponse.useSWR {
+		opts.stale = candidateResponse.overrideStaleWhileRevalidate
+	} else {
+		opts.stale = suggestedCacheWriteOptions.stale
+	}
+
+	if candidateResponse.useVary {
+		opts.vary = candidateResponse.overrideVary
+	} else {
+		opts.vary = suggestedCacheWriteOptions.vary
+	}
+
+	if candidateResponse.useSurrogate {
+		opts.surrogate = candidateResponse.overrideSurrogateKeys
+	} else {
+		opts.surrogate = suggestedCacheWriteOptions.surrogate
+	}
+
+	if candidateResponse.usePCI {
+		opts.sensitive = candidateResponse.overridePCI
+	} else {
+		opts.sensitive = suggestedCacheWriteOptions.sensitive
+	}
+
+	if candidateResponse.bodyTransform == nil {
+		if len, ok := bodyHasKnownLength(candidateResponse.abiBody); ok {
+			opts.length = len
+			opts.useLength = true
+		}
+	}
+
+	opts.flushToABI()
+
+	return storageAction, &opts, nil
+}
+
+func bodyHasKnownLength(body io.ReadCloser) (uint64, bool) {
+	if abiBody, ok := body.(*fastly.HTTPBody); ok {
+		l, err := abiBody.Length()
+		if err != nil {
+			return 0, false
+		}
+		return l, true
+	}
+
+	if lenner, ok := body.(interface{ Len() int }); ok {
+		return uint64(lenner.Len()), true
+	}
```

Comment:
Could/should this same interface approach be used for `Length()` above?

Reply:
It probably *could*, but really this was s short-hand for a type-switch on `*bytes.Buffer`, `*bytes.Reader`, `*strings.Reader` which are realistically what the body types are going to be.   As for `Length() (error, uint64)` that's not a common method -- anybody is going to implement `Len() int`.  (The `HTTPBody` type doesn't implement `Len()` because I need an error return in case the body is being streamed into the cache and we don't have a known length).

(Actually, I suppose I could rewrite this whole method to be a type-switch, but I don't want to limit the concrete types that can be passed in, since we don't elsewhere for the body.)

Reply:
Fine to leave as-is.